test/hotspot/jtreg/compiler/vectorization/TestRotateByteVector.java

```diff
@@ -0,0 +1,105 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8261022
+ * @summary Test vectorization of rotate byte
+ * @run main/othervm -XX:-TieredCompilation -Xcomp -XX:CompileCommand=compileonly,TestRotateByteVector::testRotate* -Xbatch TestRotateByteVector
+ */
+
+public class TestRotateByteVector {
+    private static final int ARRLEN = 512;
+    private static final int ITERS = 50000;
+    private static byte[] arr = new byte[ARRLEN];
+    private static byte[] rolTest = new byte[ARRLEN];
+    private static byte[] rorTest = new byte[ARRLEN];
+    private static byte[] res = new byte[ARRLEN];
+
+    public static void main(String[] args) {
+        // init
+        for (int i = 0; i < ARRLEN; i++) {
+            arr[i] = (byte) i;
+        }
+
+        System.out.println("warmup");
+        warmup();
+
+        System.out.println("Testing rotate byte...");
+        test();
+
+        System.out.println("PASSED");
+    }
+
+    static void warmup() {
+        for (int i = 0; i < ITERS; i++) {
+            testRotateLeft(arr, rolTest, i);
+            testRotateRight(arr, rorTest, i);
+        }
+    }
+
+    static void test() {
+        for (int shift = 0; shift <= 512; shift++) {
+            testRotateLeft(arr, rolTest, shift);
+            rotateLeftRes(arr, res, shift);
+            verify(rolTest, res, shift, "rol");
+
+            testRotateRight(arr, rorTest, shift);
+            rotateRightRes(arr, res, shift);
+            verify(rorTest, res, shift, "ror");
+        }
+    }
+
+    static void testRotateLeft(byte[] arr, byte[] test, int shift) {
+        for (int i = 0; i < ARRLEN; i++) {
+            test[i] = (byte) ((arr[i] << shift) | (arr[i] >>> -shift));
+        }
+    }
+
+    static void testRotateRight(byte[] arr, byte[] test, int shift) {
+        for (int i = 0; i < ARRLEN; i++) {
+            test[i] = (byte) ((arr[i] >>> shift) | (arr[i] << -shift));
+        }
+    }
+
+    static void rotateLeftRes(byte[] arr, byte[] res, int shift) {
+        for (int i = 0; i < ARRLEN; i++) {
+            res[i] = (byte) ((arr[i] << shift) | (arr[i] >>> -shift));
+        }
+    }
```

Comment:
This function is duplicated with with `testRotateLeft`.  It's not enough to verify the correctness by comparing the results of these two functions.

Reply:
Yes, `rotateLeftRes` is duplicated with  `testRotateLeft`. But compile only `testRotateLeft`,  and then compare the result between the two function. Could you give me some suggestions ? How should I modify this ?

Reply:
Two options off the top of my head:

a) With `-Xcomp`, hardcode the expected values which are calculated by interpreter or just by hand and compared with the results of C2.

b) With `-XX:-TieredCompilation`, generate the expected results if the method was not hot enough to compiled by C2. Then compared with C2 results, which can be got when the iteration count is more than 10K. Please refer to https://github.com/openjdk/jdk/pull/5403/files/71aa6ac439b67b27828e9aabe51845fa34602837#diff-d14ca09ba5fa806904a4db333037a14621cfeba81505ba369a375c57bd90c7a8 for more detail.

Option a) is much more simpler. Option b) can verify more random values.

Reply:
You can tell the compiler to only compile the tested methods by using the `-XX:CompileCommand`.